### PR TITLE
Fix RaftJournalSystem / AlluxioMaster shutdown hook deadlock

### PR DESCRIPTION
**Pull request by jenoudet**
### Description of bug
 - The `AlluxioMaster` starts the `RaftJournalSystem` using `RaftJournalSystem#startInternal`. It now holds a `synchronized` lock on `RaftJournalSystem`.
https://github.com/Alluxio/alluxio/blob/dea6a46d67640f54d133744a9b318849fad9ae5c/core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java#L748

- `startInternal` gets to
https://github.com/Alluxio/alluxio/blob/dea6a46d67640f54d133744a9b318849fad9ae5c/core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java#L768
This action fails, for whatever reason, causing `System.exit(int)` --> `Runtime.getRuntime().exit(int)` internally in [Ratis](https://github.com/apache/ratis/blob/96db7f6801738c1c1036f205d925be02e219426c/ratis-common/src/main/java/org/apache/ratis/util/ExitUtils.java#L138). 

- `Runtime.getRuntime().exit(int)` calls the shotdown hooks, including: 
https://github.com/Alluxio/alluxio/blob/dea6a46d67640f54d133744a9b318849fad9ae5c/core/server/common/src/main/java/alluxio/ProcessUtils.java#L96-L98
`process.stop()` calls `RaftJournalSystem.stopInternal()`, which cannot proceed because of the `synchronized` lock on `RaftJournalSystem.startInternal()`.

### Result
Deadlock.

### Fix
This one line disables the `System.exit(int)` call in Ratis internally in favor of an Exception being thrown. This prevents the deadlock.

### CODE REVIEW

**Comment by ggezer** [comment]
`System.getRuntime()` should be owned only while adding the hook during `stopProcessOnShutdown` call. Once the hook is registered, I wouldn't expect the hook code itself to hold `System.getRuntime()`. So I don't understand the deadlock. Is this PR resolving a reproducible deadlock?

**Comment by jenoudet** [comment]
@ggezer I've changed the description to explain the deadlock accurately. My previous explanations contained some misunderstandings.

File: core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java [hunk 1]
```diff
@@ -380,6 +380,10 @@ private synchronized void initServer() throws IOException {
     RaftServerConfigKeys.Log.Appender.setInstallSnapshotEnabled(
         properties, false);
 
+    // if left enabled, the System.exit() called by Ratis can deadlock with the AlluxioMaster
+    // process shutdown hook
+    org.apache.ratis.util.ExitUtils.disableSystemExit();
```
Comment (ggezer) [inline core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java:385]: Please update the comment here as well. For this one, it's OK to have a big comment.

**Review by ggezer — APPROVED** [review]
Thanks for the fix.

**Review by yuzhu — APPROVED** [review]
LGTM, thanks for the fix.

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Pull request merged**